### PR TITLE
Enable groups drag'n'drop to new library

### PR DESCRIPTION
**Pull request by rafatd**
Fixes https://github.com/koppor/jabref/issues/530
- [ ] Change in `CHANGELOG.md` described in a way that is understandable for the average user (if applicable)
- [ ] Tests created for changes (if applicable)
- [x] Manually tested changed features in running JabRef (always required)
- [X] Screenshots added in PR description (for UI changes)
- [X] [Checked developer's documentation](https://devdocs.jabref.org/): Is the information available and up to date? If not, I outlined it in this pull request.
- [X] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not, I created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, I submitted a pull request to the documentation repository.
Screenshot 2022-12-07 082639
Screenshot 2022-12-07 082327
Screenshot 2022-12-07 082459
Screenshot 2022-12-07 082609

### CODE REVIEW

**Comment by ThiloteE** [comment]
To ease organizational workflows I have linked the pull-request to the issue with syntax as described in https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue

> <h2 id="linking-a-pull-request-to-an-issue-using-a-keyword">Linking a pull request to an issue using a keyword</h2>
> <p>You can link a pull request to an issue by using a supported keyword 
> in the pull request's description or in a commit message. The pull 
> request <strong>must be</strong> on the default branch.</p>
> <ul>
> <li>close</li>
> <li>closes</li>
> <li>closed</li>
> <li>fix</li>
> <li>fixes</li>
> <li>fixed</li>
> <li>resolve</li>
> <li>resolves</li>
> <li>resolved</li>
> </ul>
> <p>If you use a keyword to reference a pull request comment in another 
> pull request, the pull requests will be linked. Merging the referencing 
> pull request also closes the referenced pull request.</p>
> <p>The syntax for closing keywords depends on whether the issue is in the same repository as the pull request.</p>

**Comment by Siedlerchr** [comment]
Please have a look at the checkstlye/reviewdog tests

**Comment by rafatd** [comment]
> Please have a look at the checkstlye/reviewdog tests

I have tried many times to fix these issues, but still I keep getting this error "Use a single space to separate non-whitespace characters." Can anyone please help me understand what that mean?

**Comment by Siedlerchr** [comment]
@rafatd the best is to setup JabRef's code style https://devdocs.jabref.org/getting-into-the-code/guidelines-for-setting-up-a-local-workspace.html#using-jabrefs-code-style 
Then you can just use the ide formatting

File: src/main/java/org/jabref/gui/JabRefFrame.java [hunk 1]
```diff
@@ -263,7 +267,43 @@ private void initDragAndDrop() {
                             for (Tab libraryTab : tabbedPane.getTabs()) {
```
Comment (Siedlerchr) [inline src/main/java/org/jabref/gui/JabRefFrame.java:267]: Name this variable tab

File: src/main/java/org/jabref/gui/JabRefFrame.java [hunk 2]
```diff
@@ -263,7 +267,43 @@ private void initDragAndDrop() {
...
                                 if (libraryTab.getId().equals(tab.getId()) &&
                                         !tabbedPane.getSelectionModel().getSelectedItem().equals(libraryTab)) {
-                                    ((LibraryTab) libraryTab).dropEntry(stateManager.getLocalDragboard().getBibEntries());
+                                    if (dragboard.hasContent(DragAndDropDataFormats.GROUP)) {
+                                        List<String> groupPathToSources = (List<String>) dragboard.getContent(DragAndDropDataFormats.GROUP);
+                                        if (((LibraryTab) libraryTab).getBibDatabaseContext().getMetaData().getGroups().isEmpty()) {
```
Comment (Siedlerchr) [inline src/main/java/org/jabref/gui/JabRefFrame.java:272]: I would extract this casting to LibraryTab to a new variable, then you don't have to add the casting everywhere
e.g. LibraryTab libraryTab = (LibraryTab) tab;

**Comment by Siedlerchr** [comment]
Tested your feature, looks good so far. Just one thing I noticed: Colors and icons are not transfered. Don't know if this is possible as well

File: src/main/java/org/jabref/gui/JabRefFrame.java (hunk 1 above)
Comment (rafatd) [reply src/main/java/org/jabref/gui/JabRefFrame.java:267]: With regard to this, I did not add this line to the file. And I believe naming it a tab would case issues, as another variable named tab is already predefined in the scope.

Comment (Siedlerchr) [reply src/main/java/org/jabref/gui/JabRefFrame.java:267]: Ah okay, I see. I was just thinking about to extract the casting

File: src/main/java/org/jabref/gui/JabRefFrame.java (hunk 2 above)
Comment (rafatd) [reply src/main/java/org/jabref/gui/JabRefFrame.java]: Noted, thanks for that, I will update the file accordingly. Thanks.

**Comment by rafatd** [comment]
> Tested your feature, looks good so far. Just one thing I noticed: Colors and icons are not transferred. Don't know if this is possible as well

I just noticed that. That's weird, since the colour and the icon are stored in the abstract group object, and I am deeply copying that. So I am confused as to why it's not transferring them. I will try to make more research regarding that. In the main time, I can confirm that my last commit passes all the tests. Let me know if I must change anything, or if you are happy with that.

File: src/main/java/org/jabref/gui/JabRefFrame.java (hunk 2 above)
Comment (rafatd) [reply src/main/java/org/jabref/gui/JabRefFrame.java]: I can confirm that this has been updated.

**Comment by rafatd** [comment]
> Tested your feature, looks good so far. Just one thing I noticed: Colors and icons are not transfered. Don't know if this is possible as well

Hi again, I can confirm that I fixed that, tuners the deepcopy() function only copy deep copying an explicitgroup object, which does not know anything about the colours and the  icons. So instead I choose to copy the whole GroupTreeNode. 
Let me know if that works on your end.

**Comment by ThiloteE** [comment]
I tried the pull-request with latest commits just now and found that groups were correctly copied (as far as I could see), but encountered  issues related to:

- duplicate detection
   - have multiple entries (maybe even similar entries) in the groups and drag n' drop to new library --> entry merge editor opens. Entry merge editor should probably not open, since we simply want to copy. (Although it helps to find bugs, when the entry editor opens :D)

**Comment by rafatd** [comment]
> I tried the pull-request with latest commits just now and found that groups were correctly copied (as far as I could see), but encountered issues related to:
> 
> * duplicate detection
>   
>   * have multiple entries (maybe even similar entries) in the groups and drag n' drop to new library --> entry merge editor opens. Entry merge editor should probably not open, since we simply want to copy. (Although it helps to find bugs, when the entry editor opens :D)

I believe that I cannot reproduce this error on my local machine. If you have the same entires the merge editor will be open, this had nothing to do with my feature I am assuming. It would be helpful if you can send me the exact steps you did to get that?

**Comment by HoussemNasri** [comment]
After dragging and dropping a group and then deleting it, a notification that says "The library has been modified by another program" is triggered. However, it seems to be an old bug that was not introduced in this PR. https://github.com/JabRef/jabref/issues/9064

**Comment by rafatd** [comment]
> After dragging and dropping a group and then deleting it, a notification that says "The library has been modified by another program" is triggered. However, it seems to be an old bug that was not introduced in this PR. #9064

I cannot reproduce this issue, as you say it's an old bug! does fixing the issue https://github.com/koppor/jabref/issues/530 also require fixing that?

**Comment by HoussemNasri** [comment]
You can ignore the comment. It's for documentation purposes to clarify that the bug existed before this PR. It's definitely not a requirement to fix the bug, but you're welcome to give it a try. However, keep in mind that it is not an easy catch. We had many previous attempts of fixing it but it keeps coming back.

**Comment by rafatd** [comment]
> You can ignore the comment. It's for documentation purposes to clarify that the bug existed before this PR. It's definitely not a requirement to fix the bug, but you're welcome to give it a try. However, keep in mind that it is not an easy catch. We had many previous attempts of fixing it but it keeps coming back.

Great, I am not sure that I have enough time to be working on that. 
Is there anything else you reckon needs to be changed?

File: src/main/java/org/jabref/gui/JabRefFrame.java [hunk 3]
```diff
@@ -262,7 +266,44 @@ private void initDragAndDrop() {
                             for (Tab libraryTab : tabbedPane.getTabs()) {
                                 if (libraryTab.getId().equals(tab.getId()) &&
                                         !tabbedPane.getSelectionModel().getSelectedItem().equals(libraryTab)) {
-                                    ((LibraryTab) libraryTab).dropEntry(stateManager.getLocalDragboard().getBibEntries());
+                                    LibraryTab libraryTabCasted = (LibraryTab) libraryTab;
+                                    if (dragboard.hasContent(DragAndDropDataFormats.GROUP)) {
+                                        List<String> groupPathToSources = (List<String>) dragboard.getContent(DragAndDropDataFormats.GROUP);
+                                        if (libraryTabCasted.getBibDatabaseContext().getMetaData().getGroups().isEmpty()) {
```
Comment (HoussemNasri) [inline src/main/java/org/jabref/gui/JabRefFrame.java:272]: Please consider reducing the level of nesting. Nested code is hard to follow. Physically there’s more visual scanning with the eyes. Mentally, each level of nesting requires more overhead to track functionality. All of this exhaust a reader.

A common solution to reduce nesting is by inverting if statements; You can read more about it [here](https://stackoverflow.com/questions/268132/invert-if-statement-to-reduce-nesting). In addition, you can replace for loops with Java 8 stream operations *(map-filter-flatmap)* when possible. It will make your code so much easier to follow.

File: src/main/java/org/jabref/gui/JabRefFrame.java [hunk 4]
```diff
@@ -1422,4 +1463,22 @@ private void updateTexts(UndoChangeEvent event) {
             */
         }
     }
+
+    private void copyGroupTreeNode(LibraryTab libraryTabCasted, Optional<GroupTreeNode> parent, Optional<GroupTreeNode> groupTreeNodeToCopy, List<BibEntry> allEntries) {
+        // add groupTreeNodeToCopy to the parent-- in the first run that will the source/main GroupTreeNode
+        GroupTreeNode copiedNode = parent.get().addSubgroup(groupTreeNodeToCopy.get().copyNode().getGroup());
```
Comment (HoussemNasri) [inline src/main/java/org/jabref/gui/JabRefFrame.java:1469]: Passing Optional to methods is generally a bad practice. Plus, I don't see how they're useful in this context, because *parent* and *groupTreeNodeToCopy* are being called without checking whether they exist or not.

File: src/main/java/org/jabref/gui/JabRefFrame.java [hunk 5]
```diff
@@ -262,7 +266,44 @@ private void initDragAndDrop() {
...
+                                                    .getBibDatabaseContext()
+                                                    .getEntries();
+                                            Optional<GroupTreeNode> groupTreeNodeToCopy = getCurrentLibraryTab()
+                                                    .getBibDatabaseContext()
+                                                    .getMetaData()
+                                                    .getGroups()
+                                                    .get()
+                                                    .getChildByPath(pathToSource);
+
+                                            // call the copy function
```
Comment (HoussemNasri) [inline src/main/java/org/jabref/gui/JabRefFrame.java:301]: I think you forgot this comment.

File: src/main/java/org/jabref/gui/JabRefFrame.java [hunk 6]
```diff
@@ -262,7 +266,44 @@ private void initDragAndDrop() {
                             for (Tab libraryTab : tabbedPane.getTabs()) {
                                 if (libraryTab.getId().equals(tab.getId()) &&
                                         !tabbedPane.getSelectionModel().getSelectedItem().equals(libraryTab)) {
-                                    ((LibraryTab) libraryTab).dropEntry(stateManager.getLocalDragboard().getBibEntries());
+                                    LibraryTab libraryTabCasted = (LibraryTab) libraryTab;
```
Comment (HoussemNasri) [inline src/main/java/org/jabref/gui/JabRefFrame.java:269]: Can you please assign tab and libraryTabCasted more expressive names. For instance, you can rename them to *destinationTabNode* and *destinationLibraryTab*.

File: src/main/java/org/jabref/gui/JabRefFrame.java [hunk 7]
```diff
@@ -1422,4 +1463,22 @@ private void updateTexts(UndoChangeEvent event) {
             */
         }
     }
+
+    private void copyGroupTreeNode(LibraryTab libraryTabCasted, Optional<GroupTreeNode> parent, Optional<GroupTreeNode> groupTreeNodeToCopy, List<BibEntry> allEntries) {
```
Comment (HoussemNasri) [inline src/main/java/org/jabref/gui/JabRefFrame.java:1467]: Is passing *allEntries* here really necessary?

File: src/main/java/org/jabref/gui/JabRefFrame.java [hunk 8]
```diff
@@ -262,7 +266,44 @@ private void initDragAndDrop() {
                             for (Tab libraryTab : tabbedPane.getTabs()) {
                                 if (libraryTab.getId().equals(tab.getId()) &&
                                         !tabbedPane.getSelectionModel().getSelectedItem().equals(libraryTab)) {
-                                    ((LibraryTab) libraryTab).dropEntry(stateManager.getLocalDragboard().getBibEntries());
+                                    LibraryTab libraryTabCasted = (LibraryTab) libraryTab;
+                                    if (dragboard.hasContent(DragAndDropDataFormats.GROUP)) {
+                                        List<String> groupPathToSources = (List<String>) dragboard.getContent(DragAndDropDataFormats.GROUP);
+                                        if (libraryTabCasted.getBibDatabaseContext().getMetaData().getGroups().isEmpty()) {
+                                            // a root (all entries) GroupTreeNode
+                                            GroupTreeNode rootNode = getCurrentLibraryTab().getBibDatabaseContext()
```
Comment (HoussemNasri) [inline src/main/java/org/jabref/gui/JabRefFrame.java:274]: When I saw this code, I asked root node of what. A better name would be *currentLibraryGroupRoot*.

File: src/main/java/org/jabref/gui/JabRefFrame.java [hunk 9]
```diff
@@ -262,7 +266,44 @@ private void initDragAndDrop() {
...
+                                                                                           .getGroups()
+                                                                                           .get()
+                                                                                           .copyNode();
+
+                                            // add rootNode to the Library if it does not have a root.
+                                            libraryTabCasted.getBibDatabaseContext()
+                                                                     .getMetaData()
+                                                                     .setGroups(rootNode);
+                                        }
+                                        Optional<GroupTreeNode> parentGroup = Optional.of(libraryTabCasted
```
Comment (HoussemNasri) [inline src/main/java/org/jabref/gui/JabRefFrame.java:285]: The same comment as the above. It could be renamed to *destinationLibraryGroupRoot*

File: src/main/java/org/jabref/gui/JabRefFrame.java [hunk 10]
```diff
@@ -1422,4 +1463,22 @@ private void updateTexts(UndoChangeEvent event) {
...
+        // add all entries of a groupTreeNode to the new library.
+        libraryTabCasted.dropEntry(groupTreeNodeToCopy.get().getEntriesInGroup(allEntries));
+
+        // List of all children of groupTreeNodeToCopy
+        List<GroupTreeNode> children = groupTreeNodeToCopy.get().getChildren();
+
+        if (!children.isEmpty()) {
+            // use recursion to add all subgroups of the original groupTreeNodeToCopy
+            for (GroupTreeNode child : children) {
+                copyGroupTreeNode(libraryTabCasted, Optional.of(copiedNode), Optional.of(child), allEntries);
```
Comment (HoussemNasri) [inline src/main/java/org/jabref/gui/JabRefFrame.java:1480]: An alternative way to copying groups https://github.com/JabRef/jabref/blob/e86839a85fb939920e3a3d0214aeda9a0b47ae44/src/main/java/org/jabref/gui/collab/groupchange/GroupChange.java#L44-L46

File: src/main/java/org/jabref/gui/JabRefFrame.java [hunk 11]
```diff
@@ -234,7 +236,7 @@ private void initDragAndDrop() {
                 // drag'n'drop on tabs covered dnd on tabbedPane, so dnd on tabs should contain all dnds on tabbedPane
                 tabbedPane.lookupAll(".tab").forEach(tab -> {
                     tab.setOnDragOver(tabDragEvent -> {
-                        if (DragAndDropHelper.hasBibFiles(tabDragEvent.getDragboard())) {
+                        if (DragAndDropHelper.hasBibFiles(tabDragEvent.getDragboard()) || tabDragEvent.getDragboard().hasContent(DragAndDropDataFormats.GROUP)) {
```
Comment (HoussemNasri) [inline src/main/java/org/jabref/gui/JabRefFrame.java:239]: I would use similar methods to `DragAndDropHelper#hasBibFiles` and `DragAndDropHelper.getBibFiles` for better readability.

**Review by HoussemNasri — CHANGES_REQUESTED** [review]
I tried out the feature on my local machine and it works great! However, I have a few comments about the code.

File: src/main/java/org/jabref/gui/JabRefFrame.java (hunk 5 above)
Comment (rafatd) [reply src/main/java/org/jabref/gui/JabRefFrame.java:301]: I do not fully understand what do you mean by I forgot it, can you please clarify that?

File: src/main/java/org/jabref/gui/JabRefFrame.java (hunk 6 above)
Comment (rafatd) [reply src/main/java/org/jabref/gui/JabRefFrame.java:269]: I can confirm that I will be changing the `libraryTabCasted` to `destinationLibraryTab`- thanks for that - , however, for `tab` I have not added that and as you can see it has been used by the name tab many times in this file. So I do not feel ok changing that? Can you please let me know what you think?

File: src/main/java/org/jabref/gui/JabRefFrame.java (hunk 9 above)
Comment (rafatd) [reply src/main/java/org/jabref/gui/JabRefFrame.java:285]: Will be edited, thanks.

File: src/main/java/org/jabref/gui/JabRefFrame.java (hunk 11 above)
Comment (rafatd) [reply src/main/java/org/jabref/gui/JabRefFrame.java:239]: These methods do not support checking having groups, so I will create some methods in the `DragAndDropHelper `class. Thanks.

File: src/main/java/org/jabref/gui/JabRefFrame.java (hunk 4 above)
Comment (rafatd) [reply src/main/java/org/jabref/gui/JabRefFrame.java:1469]: Noted, thanks.

File: src/main/java/org/jabref/gui/JabRefFrame.java (hunk 5 above)
Comment (HoussemNasri) [reply src/main/java/org/jabref/gui/JabRefFrame.java:301]: The comment doesn't add any value to the code. Anyone who knows Java should understand that the next line is calling the copy method :^). Generally you want to use comments to explain things that you can't express fluently in the code through the available syntax and good variable naming practice.

The reason behind avoiding comments when possible, is that comments tend to go out of date without any warning. For instance, if you change a method name, you will get compilations errors in all places where the method was used. You won't get that with comments, which can risk passing the wrong information to code readers.

File: src/main/java/org/jabref/gui/JabRefFrame.java (hunk 7 above)
Comment (HoussemNasri) [reply src/main/java/org/jabref/gui/JabRefFrame.java:1467]: But still, you don't have to pass it as an argument. You can create a local variable inside the copy method and have it hold all entries.

File: src/main/java/org/jabref/gui/JabRefFrame.java (hunk 6 above)
Comment (HoussemNasri) [reply src/main/java/org/jabref/gui/JabRefFrame.java:269]: Yeah, you can go ahead and rename it. I looked into the other code and destinationLibraryTabNode *(or destinationTabNode)* fits well in there too.

File: src/main/java/org/jabref/gui/JabRefFrame.java [hunk 12]
```diff
@@ -203,6 +206,7 @@ public JabRefFrame(Stage mainStage) {
         });
     }
 
+    @SuppressWarnings("checkstyle:SingleSpaceSeparator")
```
Comment (Siedlerchr) [inline src/main/java/org/jabref/gui/JabRefFrame.java:209]: This is not the right way to handle checkstlye issues..

Comment (rafatd) [reply src/main/java/org/jabref/gui/JabRefFrame.java:209]: I do not belive that I have added this.

**Review by Siedlerchr — APPROVED** [review]
